File: server-go/internal/domain/service/profile_service.go
package service

import (
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/aitrack/server/internal/domain/model"
)

// RawRecord holds the columns needed to compute a developer profile.
// It is populated by the persistence adapter and consumed by pure analytics.
type RawRecord struct {
	Tool          string
	FilePath      string
	AddedLines    int64
	RemovedLines  int64
	DiffHunk      string
	Timestamp     string // Unix epoch as text
	Status        string
	PromptSummary *string
}

// ProfileService computes developer profiles from raw edit records.
// It is pure: it never touches a database or the network.
type ProfileService struct{}

// NewProfileService constructs a ProfileService.
func NewProfileService() *ProfileService { return &ProfileService{} }

// BuildProfile aggregates raw records into a ProfileDto.
// owner and tokenKey identify the subject; now anchors the time-window math.
func (p *ProfileService) BuildProfile(tokenKey, owner string, records []RawRecord, now time.Time) *model.ProfileDto {
	now = now.UTC()
	dto := &model.ProfileDto{
		TokenKey:    tokenKey,
		Owner:       owner,
		GeneratedAt: now.Format(time.RFC3339),
		Languages:   make(map[string]int64),
		Tools:       make(map[string]int64),
	}

	if len(records) == 0 {
		dto.Frequency = &model.FrequencyStats{DailyTrend: []model.DayCount{}}
		dto.Depth = &model.DepthStats{}
		dto.PromptPatterns = ComputePromptPatterns(records)
		return dto
	}

	var (
		totalAdded   int64
		totalRemoved int64
		lineTotals   []int64
		minEpoch     int64 = -1
		maxEpoch     int64 = -1

		count30d int64
		count12w int64

		cutoff30d = now.Unix() - 30*86400
		cutoff12w = now.Unix() - 84*86400

		dayBuckets = make(map[string]int64)
	)

	for _, rec := range records {
		totalAdded += rec.AddedLines
		totalRemoved += rec.RemovedLines

		lineTotal := rec.AddedLines + rec.RemovedLines
		lineTotals = append(lineTotals, lineTotal)

		dto.Tools[rec.Tool]++
		dto.Languages[DetectLanguage(rec.FilePath)]++

		epochTs, parseErr := strconv.ParseInt(rec.Timestamp, 10, 64)
		if parseErr == nil {
			if minEpoch < 0 || epochTs < minEpoch {
				minEpoch = epochTs
			}
			if epochTs > maxEpoch {
				maxEpoch = epochTs
			}
			if epochTs >= cutoff30d {
				count30d++
				day := time.Unix(epochTs, 0).UTC().Format("2006-01-02")
				dayBuckets[day]++
			}
			if epochTs >= cutoff12w {
				count12w++
			}
		}
	}

	dto.TotalEdits = int64(len(records))
	dto.TotalAddedLines = totalAdded
	dto.TotalRemovedLines = totalRemoved

	if minEpoch >= 0 {
		s := time.Unix(minEpoch, 0).UTC().Format(time.RFC3339)
		dto.FirstSeen = &s
	}
	if maxEpoch >= 0 {
		s := time.Unix(maxEpoch, 0).UTC().Format(time.RFC3339)
		dto.LastSeen = &s
	}

	dailyTrend := make([]model.DayCount, 0, len(dayBuckets))
	for day, cnt := range dayBuckets {
		dailyTrend = append(dailyTrend, model.DayCount{Date: day, Count: cnt})
	}
	sort.Slice(dailyTrend, func(i, j int) bool { return dailyTrend[i].Date < dailyTrend[j].Date })
	dto.Frequency = &model.FrequencyStats{
		DailyAvg30d:  float64(count30d) / 30.0,
		WeeklyAvg12w: float64(count12w) / 12.0,
		DailyTrend:   dailyTrend,
	}

	sort.Slice(lineTotals, func(i, j int) bool { return lineTotals[i] < lineTotals[j] })
	n := int64(len(lineTotals))
	var sum, smallCount, mediumCount, largeCount int64
	for _, v := range lineTotals {
		sum += v
		switch {
		case v < 10:
			smallCount++
		case v <= 100:
			mediumCount++
		default:
			largeCount++
		}
	}
	dto.Depth = &model.DepthStats{
		AvgLines:       float64(sum) / float64(n),
		P50Lines:       lineTotals[n/2],
		P90Lines:       lineTotals[int(float64(n)*0.9)],
		SmallCount:     smallCount,
		MediumCount:    mediumCount,
		LargeCount:     largeCount,
		CommentDensity: ComputeCommentDensity(records),
	}

	dto.PromptPatterns = ComputePromptPatterns(records)
	return dto
}

// DetectLanguage maps a file path to a language name based on its extension.
func DetectLanguage(filePath string) string {
	if filePath == "" {
		return "Other"
	}
	ext := strings.ToLower(filepath.Ext(filePath))
	switch ext {
	case ".py":
		return "Python"
	case ".ts", ".tsx":
		return "TypeScript"
	case ".js", ".jsx":
		return "JavaScript"
	case ".java":
		return "Java"
	case ".go":
		return "Go"
	case ".rs":
		return "Rust"
	case ".cpp", ".cc", ".cxx", ".c":
		return "C/C++"
	case ".cs":
		return "C#"
	case ".rb":
		return "Ruby"
	case ".php":
		return "PHP"
	case ".swift":
		return "Swift"
	case ".kt", ".kts":
		return "Kotlin"
	case ".scala":
		return "Scala"
	case ".vue":
		return "Vue"
	case ".html", ".htm":
		return "HTML"
	case ".css", ".scss", ".sass", ".less":
		return "CSS"
	case ".sql":
		return "SQL"
	case ".sh", ".bash", ".zsh":
		return "Shell"
	case ".yaml", ".yml":
		return "YAML"
	case ".json":
		return "JSON"
	case ".xml":
		return "XML"
	case ".md", ".rst", ".txt":
		return "Docs"
	default:
		return "Other"
	}
}

// ComputeCommentDensity calculates the ratio of comment lines to total added
// lines across all records' diff hunks.
func ComputeCommentDensity(records []RawRecord) float64 {
	var totalAdded, commentLines int64
	for _, r := range records {
		if r.DiffHunk == "" {
			continue
		}
		for _, line := range strings.Split(r.DiffHunk, "\n") {
			if !strings.HasPrefix(line, "+") || strings.HasPrefix(line, "+++") {
				continue
			}
			totalAdded++
			trimmed := strings.TrimSpace(line[1:]) // strip leading +
			if strings.HasPrefix(trimmed, "//") ||
				strings.HasPrefix(trimmed, "#") ||
				strings.HasPrefix(trimmed, "/*") ||
				strings.HasPrefix(trimmed, "* ") ||
				strings.HasPrefix(trimmed, "*/") ||
				strings.HasPrefix(trimmed, "/**") ||
				strings.HasPrefix(trimmed, `"""`) ||
				strings.HasPrefix(trimmed, `'''`) ||
				strings.HasPrefix(trimmed, "--") ||
				strings.HasPrefix(trimmed, "<!--") {
				commentLines++
			}
		}
	}
	if totalAdded == 0 {
		return 0.0
	}
	return float64(commentLines) / float64(totalAdded)
}

// Percentile returns the value at the given percentile (0.0–1.0) of a sorted slice.
// The slice must be sorted in ascending order.
func Percentile(sorted []int64, p float64) int64 {
	if len(sorted) == 0 {
		return 0
	}
	idx := int(float64(len(sorted)) * p)
	if idx >= len(sorted) {
		idx = len(sorted) - 1
	}
	return sorted[idx]
}

File: server-java/src/main/java/com/aitrack/server/service/ProfileService.java
package com.aitrack.server.service;

import com.aitrack.server.dto.ProfileDto;
import com.aitrack.server.dto.ProfileDto.DayCount;
import com.aitrack.server.dto.ProfileDto.DepthStats;
import com.aitrack.server.dto.ProfileDto.FrequencyStats;
import com.aitrack.server.entity.EditRecordEntity;
import com.aitrack.server.entity.EditRecordEntity.RecordStatus;
import com.aitrack.server.entity.TokenEntity;
import com.aitrack.server.repository.EditRecordRepository;
import com.aitrack.server.repository.TokenRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.util.*;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class ProfileService {

    private final EditRecordRepository editRecordRepo;
    private final TokenRepository tokenRepo;

    /**
     * Compute a developer AI usage profile for the given tokenKey.
     *
     * @param tokenKey the 16-character token key (aitrack format)
     * @return Optional.empty() when neither records nor an active token exist; otherwise Optional.of(profile)
     */
    public Optional<ProfileDto> computeProfile(String tokenKey) {
        List<EditRecordEntity> records =
                editRecordRepo.findByTokenKeyAndStatusNot(tokenKey, RecordStatus.REJECTED);
        Optional<TokenEntity> tokenOpt = tokenRepo.findByTokenKeyAndActiveTrue(tokenKey);

        // 404: no records AND no active token
        if (records.isEmpty() && tokenOpt.isEmpty()) {
            return Optional.empty();
        }

        ProfileDto profile = new ProfileDto();
        profile.setTokenKey(tokenKey);
        profile.setGeneratedAt(Instant.now().toString());

        // Owner from token entity
        tokenOpt.ifPresent(t -> profile.setOwner(t.getOwner()));

        if (records.isEmpty()) {
            // Token exists but no records — return zero-value profile
            profile.setTotalEdits(0);
            profile.setTotalAddedLines(0);
            profile.setTotalRemovedLines(0);
            return Optional.of(profile);
        }

        // ------------------------------------------------------------------
        // Basic aggregates
        // ------------------------------------------------------------------
        profile.setTotalEdits(records.size());
        profile.setTotalAddedLines(records.stream().mapToLong(EditRecordEntity::getAddedLines).sum());
        profile.setTotalRemovedLines(records.stream().mapToLong(EditRecordEntity::getRemovedLines).sum());

        // ------------------------------------------------------------------
        // Time bounds (first_seen / last_seen)
        // ------------------------------------------------------------------
        List<Instant> instants = new ArrayList<>();
        for (EditRecordEntity e : records) {
            instants.add(parseTimestamp(e));
        }
        instants.sort(Comparator.naturalOrder());
        if (!instants.isEmpty()) {
            profile.setFirstSeen(instants.get(0).toString());
            profile.setLastSeen(instants.get(instants.size() - 1).toString());
        }

        // ------------------------------------------------------------------
        // Frequency stats
        // ------------------------------------------------------------------
        profile.setFrequency(computeFrequency(records));

        // ------------------------------------------------------------------
        // Depth stats
        // ------------------------------------------------------------------
        profile.setDepth(computeDepth(records));

        // ------------------------------------------------------------------
        // Language breakdown
        // ------------------------------------------------------------------
        profile.setLanguages(computeLanguages(records));

        // ------------------------------------------------------------------
        // Comment density (set into DepthStats after depth is computed)
        // ------------------------------------------------------------------
        ProfileDto.DepthStats depthStats = profile.getDepth();
        if (depthStats != null) {
            depthStats.setCommentDensity(computeCommentDensity(records));
        }

        // ------------------------------------------------------------------
        // Tools breakdown (skip null tools)
        // ------------------------------------------------------------------
        Map<String, Long> tools = records.stream()
                .filter(e -> e.getTool() != null)
                .collect(Collectors.groupingBy(
                        EditRecordEntity::getTool,
                        Collectors.counting()
                ));
        profile.setTools(tools.isEmpty() ? null : tools);

        // ------------------------------------------------------------------
        // Prompt patterns
        // ------------------------------------------------------------------
        profile.setPromptPatterns(computePromptPatterns(records));

        return Optional.of(profile);
    }

    // -------------------------------------------------------------------------
    // Frequency computation
    // -------------------------------------------------------------------------

    private FrequencyStats computeFrequency(List<EditRecordEntity> records) {
        Instant now = Instant.now();
        Instant cutoff30d = now.minusSeconds(30L * 24 * 3600);
        Instant cutoff84d = now.minusSeconds(84L * 24 * 3600);

        long count30d = 0;
        long count84d = 0;
        Map<LocalDate, Long> dailyCounts = new TreeMap<>();

        for (EditRecordEntity e : records) {
            Instant ts = parseTimestamp(e);
            if (!ts.isBefore(cutoff30d)) {
                count30d++;
                LocalDate day = ts.atZone(ZoneOffset.UTC).toLocalDate();
                dailyCounts.merge(day, 1L, Long::sum);
            }
            if (!ts.isBefore(cutoff84d)) {
                count84d++;
            }
        }

        FrequencyStats freq = new FrequencyStats();
        freq.setDailyAvg30d(count30d / 30.0);
        freq.setWeeklyAvg12w(count84d / 12.0);

        List<DayCount> trend = dailyCounts.entrySet().stream()
                .map(entry -> new DayCount(entry.getKey().toString(), entry.getValue()))
                .collect(Collectors.toList());
        freq.setDailyTrend(trend);

        return freq;
    }

    // -------------------------------------------------------------------------
    // Depth computation
    // -------------------------------------------------------------------------

    private DepthStats computeDepth(List<EditRecordEntity> records) {
        List<Long> totals = records.stream()
                .map(e -> e.getAddedLines() + e.getRemovedLines())
                .sorted()
                .collect(Collectors.toList());

        DepthStats depth = new DepthStats();

        if (totals.isEmpty()) {
            return depth;
        }

        double avg = totals.stream().mapToLong(Long::longValue).average().orElse(0.0);
        depth.setAvgLines(avg);
        depth.setP50Lines(totals.get(totals.size() / 2));
        depth.setP90Lines(totals.get((int) (totals.size() * 0.9)));
        depth.setSmallCount(totals.stream().filter(t -> t < 10).count());
        depth.setMediumCount(totals.stream().filter(t -> t >= 10 && t <= 100).count());
        depth.setLargeCount(totals.stream().filter(t -> t > 100).count());

        return depth;
    }

    // -------------------------------------------------------------------------
    // Language detection
    // -------------------------------------------------------------------------

    /**
     * Detects the programming language from a file path based on its extension.
     *
     * @param filePath the file path to examine (may be null/blank)
     * @return language name string, e.g. "Python", "TypeScript", or "Other"
     */
    String detectLanguage(String filePath) {
        if (filePath == null || filePath.isBlank()) {
            return "Other";
        }
        String lower = filePath.toLowerCase();

        if (lower.endsWith(".py"))                                                  return "Python";
        if (lower.endsWith(".ts") || lower.endsWith(".tsx"))                       return "TypeScript";
        if (lower.endsWith(".js") || lower.endsWith(".jsx"))                       return "JavaScript";
        if (lower.endsWith(".java"))                                                return "Java";
        if (lower.endsWith(".go"))                                                  return "Go";
        if (lower.endsWith(".rs"))                                                  return "Rust";
        if (lower.endsWith(".cpp") || lower.endsWith(".cc") || lower.endsWith(".cxx") || lower.endsWith(".c"))
                                                                                    return "C/C++";
        if (lower.endsWith(".cs"))                                                  return "C#";
        if (lower.endsWith(".rb"))                                                  return "Ruby";
        if (lower.endsWith(".php"))                                                 return "PHP";
        if (lower.endsWith(".swift"))                                               return "Swift";
        if (lower.endsWith(".kt") || lower.endsWith(".kts"))                       return "Kotlin";
        if (lower.endsWith(".scala"))                                               return "Scala";
        if (lower.endsWith(".vue"))                                                 return "Vue";
        if (lower.endsWith(".html") || lower.endsWith(".htm"))                     return "HTML";
        if (lower.endsWith(".css") || lower.endsWith(".scss") || lower.endsWith(".sass") || lower.endsWith(".less"))
                                                                                    return "CSS";
        if (lower.endsWith(".sql"))                                                 return "SQL";
        if (lower.endsWith(".sh") || lower.endsWith(".bash") || lower.endsWith(".zsh"))
                                                                                    return "Shell";
        if (lower.endsWith(".yaml") || lower.endsWith(".yml"))                     return "YAML";
        if (lower.endsWith(".json"))                                                return "JSON";
        if (lower.endsWith(".xml"))                                                 return "XML";
        if (lower.endsWith(".md") || lower.endsWith(".rst") || lower.endsWith(".txt"))
                                                                                    return "Docs";

        return "Other";
    }

    /**
     * Groups records by detected language and returns a count per language.
     *
     * @param records non-null list of edit records
     * @return map from language name to occurrence count
     */
    private Map<String, Long> computeLanguages(List<EditRecordEntity> records) {
        return records.stream()
                .collect(Collectors.groupingBy(
                        e -> detectLanguage(e.getFilePath()),
                        Collectors.counting()
                ));
    }

    /**
     * Computes the comment density across all records.
     * <p>
     * For each record the {@code diffHunk} field is inspected. Lines that begin
     * with {@code +} (added lines in unified-diff format) are counted as total
     * added lines; among those, lines whose content (after stripping the leading
     * {@code +}) starts with a known comment prefix are counted as comment lines.
     *
     * @param records non-null list of edit records
     * @return ratio of comment lines to total added lines, or 0.0 if no added lines
     */
    double computeCommentDensity(List<EditRecordEntity> records) {
        long totalAdded = 0;
        long commentLines = 0;

        for (EditRecordEntity record : records) {
            String hunk = record.getDiffHunk();
            if (hunk == null) {
                continue;
            }
            for (String line : hunk.split("\n")) {
                if (!line.startsWith("+")) {
                    continue;
                }
                totalAdded++;
                String content = line.substring(1).stripLeading();
                if (content.startsWith("//")
                        || content.startsWith("#")
                        || content.startsWith("/*")
                        || content.startsWith("* ")
                        || content.startsWith("*/")
                        || content.startsWith("/**")
                        || content.startsWith("\"\"\"")
                        || content.startsWith("'''")
                        || content.startsWith("--")
                        || content.startsWith("<!--")) {
                    commentLines++;
                }
            }
        }

        if (totalAdded == 0) {
            return 0.0;
        }
        return (double) commentLines / totalAdded;
    }

    // -------------------------------------------------------------------------
    // Prompt pattern classification
    // -------------------------------------------------------------------------

    private Map<String, Long> computePromptPatterns(List<EditRecordEntity> records) {
        Map<String, Long> patterns = new java.util.LinkedHashMap<>();
        patterns.put("generate", 0L);
        patterns.put("fix_debug", 0L);
        patterns.put("refactor", 0L);
        patterns.put("explain", 0L);
        patterns.put("test", 0L);
        patterns.put("other", 0L);

        for (EditRecordEntity r : records) {
            String ps = r.getPromptSummary();
            if (ps == null || ps.isBlank()) continue;
            String lower = ps.toLowerCase();
            if (lower.matches(".*(generate|create|write|implement|add).*")) {
                patterns.merge("generate", 1L, Long::sum);
            } else if (lower.matches(".*(fix|debug|error|bug|broken|failing).*")) {
                patterns.merge("fix_debug", 1L, Long::sum);
            } else if (lower.matches(".*(refactor|clean|improve|reorganize|rename).*")) {
                patterns.merge("refactor", 1L, Long::sum);
            } else if (lower.matches(".*(explain|what|how|why|understand|describe).*")) {
                patterns.merge("explain", 1L, Long::sum);
            } else if (lower.matches(".*(test|spec|mock|assert|verify).*")) {
                patterns.merge("test", 1L, Long::sum);
            } else {
                patterns.merge("other", 1L, Long::sum);
            }
        }
        return patterns;
    }

    // -------------------------------------------------------------------------
    // Timestamp parsing
    // -------------------------------------------------------------------------

    /**
     * Parses the entity's timestamp field (expected to be epoch-seconds as a Long string).
     * Falls back to receivedAt if parsing fails.
     */
    private Instant parseTimestamp(EditRecordEntity e) {
        try {
            long epochSec = Long.parseLong(e.getTimestamp());
            return Instant.ofEpochSecond(epochSec);
        } catch (NumberFormatException | NullPointerException ex) {
            log.debug("Failed to parse timestamp '{}' for record id={}, falling back to receivedAt",
                    e.getTimestamp(), e.getId());
            return e.getReceivedAt();
        }
    }
}
